File: projects/08/vm_translator/command.py
from .vm import CMD_ADD
from .vm import CMD_SUB
from .vm import CMD_NEG
from .vm import CMD_AND
from .vm import CMD_OR
from .vm import CMD_NOT
from .vm import CMD_EQ
from .vm import CMD_LT
from .vm import CMD_GT
from .vm import CMD_LABEL
from .vm import CMD_GOTO
from .vm import CMD_IF_GOTO
from .vm import FRAME_ADDRS


class Command:

    def __init__(self, operation, **kwargs):
        self.operation = operation
        self.__dict__.update(**kwargs)
        self.assembly = []

    def __repr__(self):
        return '\n'.join(self.assembly) + '\n'


class MathCommand(Command):

    ops = {
        CMD_ADD: {'asm': ['M=M+D'], 'n_args': 2},
        CMD_SUB: {'asm': ['M=M-D'], 'n_args': 2},
        CMD_NEG: {'asm':  ['M=-M'], 'n_args': 1},
        CMD_AND: {'asm': ['M=M&D'], 'n_args': 2},
        CMD_OR:  {'asm': ['M=M|D'], 'n_args': 2},
        CMD_NOT: {'asm':  ['M=!M'], 'n_args': 1},
    }

    def __init__(self, operation, *args):
        super().__init__(operation)
        self.operation = self.__class__.ops[operation]['asm']
        self.n_args = self.__class__.ops[operation]['n_args']

    def translate(self):
        if self.n_args == 1:
            self.assembly += ['@SP', 'A=M-1']
        else:
            self.assembly += ['@SP', 'AM=M-1', 'D=M', 'A=A-1']
        self.assembly += self.operation


class ComparisonCommand(Command):

    ops = {CMD_EQ: ['D;JEQ'], CMD_LT: ['D;JLT'], CMD_GT: ['D;JGT']}

    def comp_incrementer():
        i = 0
        while True:
            yield i
            i += 1

    comp_counter = comp_incrementer()

    def __init__(self, operation, *args):
        super().__init__(operation)
        self.operation = self.__class__.ops[self.operation]

    def translate(self):
        comp_count = next(self.__class__.comp_counter)
        self.assembly += ['@SP', 'AM=M-1', 'D=M']
        self.assembly += ['@SP', 'A=M-1']
        self.assembly += ['D=M-D']
        self.assembly += [f'@COMP_JUMP.{comp_count}']
        self.assembly += self.operation
        self.assembly += ['@SP', 'A=M-1', 'M=0']
        self.assembly += [f'@CONTINUE_JUMP.{comp_count}']
        self.assembly += ['0;JMP']
        self.assembly += [f'(COMP_JUMP.{comp_count})']
        self.assembly += ['@SP', 'A=M-1', 'M=-1']
        self.assembly += [f'(CONTINUE_JUMP.{comp_count})']


class MemoryCommand(Command):

    def __init__(self, operation, filename, segment, index):
        kwargs = {'filename': filename, 'segment': segment, 'index': index}
        super().__init__(operation, **kwargs)
        self.seg_addr = self.get_segment_address()

    def get_segment_address(self):
        i = int(self.index)
        if self.segment in FRAME_ADDRS:
            address = FRAME_ADDRS[self.segment]
            if i == 0:
                return [address, 'A=M']
            else:
                return [address, 'D=M', f'@{i}', 'A=D+A']
        elif self.segment == 'pointer':
            assert 0 <= i <= 1, f'Invalid index for pointer: {i}'
            return [f'@R{3 + i}']
        elif self.segment == 'temp':
            assert 0 <= i <= 7, f'Invalid index for temp: {i}'
            return [f'@R{5 + i}']
        elif self.segment == 'static':
            return [f'@{self.filename}.{i}']
        elif self.segment == 'constant':
            return [f'@{i}']
        else:
            raise SystemExit(f'Invalid segment: {self.segment}')


class PopCommand(MemoryCommand):

    def __init__(self, operation, filename, segment, index):
        super().__init__(operation, filename, segment, index)

    def translate(self):
        if len(self.seg_addr) > 1:
            self.assembly += self.seg_addr
            self.assembly += ['D=A']
            self.assembly += ['@R13', 'M=D']
            self.assembly += ['@SP', 'AM=M-1', 'D=M']
            self.assembly += ['@R13', 'A=M', 'M=D']
        else:
            self.assembly += ['@SP', 'AM=M-1', 'D=M']
            self.assembly += self.seg_addr
            self.assembly += ['M=D']


class PushCommand(MemoryCommand):

    def __init__(self, operation, filename, segment, index):
        super().__init__(operation, filename, segment, index)

    def translate(self):
        self.assembly += self.seg_addr
        self.assembly += ['D=A'] if self.segment == 'constant' else ['D=M']
        self.assembly += ['@SP', 'M=M+1', 'A=M-1', 'M=D']


class ProgramFlowCommand(Command):

    formatters = {
        CMD_LABEL: '({}${})',
        CMD_GOTO: '@{}${}',
        CMD_IF_GOTO: '@{}${}',
    }

    def __init__(self, operation, filename, label, *args):
        super().__init__(operation, filename=filename, label=label)
        unformatted = self.__class__.formatters[self.operation]
        self.formatted_label = [unformatted.format(self.filename, self.label)]

    def translate(self):
        if self.operation == CMD_IF_GOTO:
            self.assembly += ['@SP', 'AM=M-1', 'D=M']
        self.assembly += self.formatted_label
        if self.operation == CMD_IF_GOTO:
            self.assembly += ['D;JNE']
        elif self.operation == CMD_GOTO:
            self.assembly += ['0;JMP']


class FunctionCommand(Command):

    def __init__(self, operation, filename, function_name, n):
        super().__init__(operation, function_name=function_name, n=n)

    def translate(self):
        n = int(self.n)
        self.assembly += [f'({self.function_name})']
        self.assembly += ['D=0', '@SP', 'M=M+1', 'A=M-1', 'M=D'] * n


class CallCommand(Command):

    def call_incrementer():
        i = 0
        while True:
            yield i
            i += 1

    call_count = call_incrementer()

    def __init__(self, operation, filename, function_name, n):
        super().__init__(operation, function_name=function_name, n=n)

    def translate(self):
        n = int(self.n)
        count = next(self.__class__.call_count)
        return_label = self.function_name + f'_RETURN.{count}'
        self.assembly += [f'@{return_label}']
        self.assembly += ['D=A']
        self.assembly += ['@SP', 'M=M+1', 'A=M-1', 'M=D']
        for addr in FRAME_ADDRS.values():
            self.assembly += [addr]
            self.assembly += ['D=M']
            self.assembly += ['@SP', 'M=M+1', 'A=M-1', 'M=D']

        self.assembly += ['@SP', 'D=M', '@LCL', 'M=D']
        self.assembly += [f'@{n + 5}', 'D=D-A', '@ARG', 'M=D']

        self.assembly += [f'@{self.function_name}']
        self.assembly += ['0;JMP']
        self.assembly += [f'({return_label})']


class ReturnCommand(Command):

    def __init__(self, operation, *args):
        super().__init__(operation)

    def translate(self):
        tmp_frame = 'R13'
        tmp_ret = 'R14'

        self.assembly += ['@LCL', 'D=M', f'@{tmp_frame}', 'M=D']
        self.assembly += [f'@{tmp_frame}', 'D=M']
        self.assembly += ['@5', 'A=D-A', 'D=M']
        self.assembly += [f'@{tmp_ret}', 'M=D']
        self.assembly += ['@SP', 'AM=M-1', 'D=M']
        self.assembly += ['@ARG', 'A=M', 'M=D']
        self.assembly += ['@ARG', 'D=M', '@SP', 'M=D+1']
        addrs = list(FRAME_ADDRS.values())
        addrs.reverse()
        for i, addr in enumerate(addrs):
            self.assembly += [f'@{tmp_frame}', 'D=M']
            self.assembly += [f'@{i + 1}', 'A=D-A', 'D=M']
            self.assembly += [addr, 'M=D']

        self.assembly += [f'@{tmp_ret}', 'A=M', '0;JMP']

File: projects/08/vm_translator/src/command.rs
pub struct MathCommand<'a> {
    pub asm_op: &'a str,
    pub n_args: usize,
}

impl<'a> MathCommand<'a> {
    pub fn new(cmd: String) -> Self {
        match cmd.as_str() {
            "add" => Self { asm_op: "M=M+D", n_args: 2},
            "sub" => Self { asm_op: "M=M-D", n_args: 2},
            "neg" => Self { asm_op: "M=-M",  n_args: 1},
            "and" => Self { asm_op: "M=M&D", n_args: 2},
            "or"  => Self { asm_op: "M=M|D", n_args: 2},
            "not" => Self { asm_op: "M=!M",  n_args: 1},
            _ => panic!("Unknown math command: {}", cmd),
        }
    }

    pub fn asm_op(&self) -> String {
        self.asm_op.to_string()
    }

    pub fn n_args(&self) -> usize {
        self.n_args
    }
}

pub struct ComparisonCommand<'a> {
    asm_op: &'a str,
}

impl<'a> ComparisonCommand<'a> {
    pub fn new(cmd: String) -> Self {
        let asm_op = match cmd.as_str() {
            "eq" => "D;JEQ",
            "lt" => "D;JLT",
            "gt" => "D;JGT",
            _ => panic!("Unknown comparison command: {}", cmd),
        };
        Self { asm_op }
    }

    pub fn asm_op(&self) -> String {
        self.asm_op.to_string()
    }
}

pub struct PushCommand {
    segment: String,
    index: usize,
}

impl PushCommand {
    pub fn new(segment: String, index: usize) -> Self {
        Self { segment, index }
    }

    pub fn segment(&self) -> &String {
        &self.segment
    }

    pub fn index(&self) -> usize {
        self.index
    }
}

pub struct PopCommand {
    segment: String,
    index: usize,
}

impl PopCommand {
    pub fn new(segment: String, index: usize) -> Self {
        Self { segment, index }
    }

    pub fn segment(&self) -> &String {
        &self.segment
    }

    pub fn index(&self) -> usize {
        self.index
    }
}

pub struct ProgramFlowCommand {
    cmd: String,
    label: String,
}

impl ProgramFlowCommand {
    pub fn new(cmd: String, label: String) -> Self {
        Self { cmd, label }
    }

    pub fn command(&self) -> &String {
        &self.cmd
    }

    pub fn label(&self) -> &String {
        &self.label
    }
}

pub struct FunctionDefCommand {
    function_name: String,
    n_locals: usize,
}

impl FunctionDefCommand {
    pub fn new(function_name: String, n_locals: usize) -> Self {
        Self { function_name, n_locals }
    }

    pub fn function_name(&self) -> &String {
        &self.function_name
    }

    pub fn n_locals(&self) -> usize {
        self.n_locals
    }
}

pub struct CallCommand {
    function_name: String,
    n_args: usize,
}

impl CallCommand {
    pub fn new(function_name: String, n_args: usize) -> Self {
        Self { function_name, n_args }
    }

    pub fn function_name(&self) -> &String {
        &self.function_name
    }

    pub fn n_args(&self) -> usize {
        self.n_args
    }
}

pub struct ReturnCommand;

pub enum CommandType<'a> {
    Math { command: MathCommand<'a> },
    Comparison { command: ComparisonCommand<'a> },
    Push { command: PushCommand },
    Pop { command: PopCommand },
    ProgramFlow { command: ProgramFlowCommand },
    FunctionDef { command: FunctionDefCommand },
    Call { command: CallCommand },
    Return,
}

pub struct Command<'a> {
    pub command_type: CommandType<'a>,
}

impl<'a> Command<'a> {
    pub fn new(c: String, arg1: Option<String>, arg2: Option<String>) -> Self {
        match c.as_str() {
            "add" | "sub" | "neg" | "and" | "or" | "not" => {
                let cmd = MathCommand::new(c);
                Self { command_type: CommandType::Math { command: cmd } }
            },
            "eq" | "lt" | "gt" => {
                let cmd = ComparisonCommand::new(c);
                Self { command_type: CommandType::Comparison { command: cmd } }
            },
            "push" | "pop" => {
                let segment = match arg1 {
                    Some(s) => s,
                    None => panic!("No segment given for push command"),
                };
                let index = match arg2 {
                    Some(i) => match i.parse() {
                        Ok(idx) => idx,
                        Err(_) => panic!("Invalid index for push command"),
                    },
                    None => panic!("No index given for push command"),
                };
                match c.as_str() {
                    "push" => {
                        let cmd = PushCommand::new(segment, index);
                        Self { command_type: CommandType::Push { command: cmd } }
                    }
                    "pop" => {
                        let cmd = PopCommand::new(segment, index);
                        Self { command_type: CommandType::Pop { command: cmd } }
                    }
                    _ => panic!("Unknown push/pop command: {}", c),
                }
            },
            "label" | "goto" | "if-goto" => {
                let label = match arg1 {
                    Some(arg) => arg,
                    None => panic!("No label given for program flow command"),
                };
                let cmd = ProgramFlowCommand::new(c, label);
                Self { command_type: CommandType::ProgramFlow { command: cmd } }
            },
            "function" => {
                let function_name = match arg1 {
                    Some(name) => name,
                    None => panic!("No function name given for function def"),
                };
                let n_locals = match arg2 {
                    Some(n) => match n.parse() {
                        Ok(n) => n,
                        Err(_) => panic!("Invalid n_locals for function"),
                    },
                    None => panic!("No n locals given for function def"),
                };
                let cmd = FunctionDefCommand::new(function_name, n_locals);
                Self { command_type: CommandType::FunctionDef { command: cmd } }
            },
            "call" => {
                let function_name = match arg1 {
                    Some(name) => name,
                    None => panic!("No function name given for call"),
                };
                let n_args = match arg2 {
                    Some(n) => match n.parse() {
                        Ok(n) => n,
                        Err(_) => panic!("Invalid n_args for call"),
                    },
                    None => panic!("No n_args given for call"),
                };
                let cmd = CallCommand::new(function_name, n_args);
                Self { command_type: CommandType::Call { command: cmd } }
            },
            "return" => Self { command_type: CommandType::Return },
            _ => panic!("Unknown command: {}", c),
        }
    }
}
